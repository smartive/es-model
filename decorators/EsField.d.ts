import { Field } from '../models/Mapping';
export declare const EsField: (data: Field) => PropertyDecorator;
