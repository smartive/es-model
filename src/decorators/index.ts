export * from './EsField';
export * from './EsProperty';
export * from './EsMapping';
