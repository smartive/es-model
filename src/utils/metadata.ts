export const KEY = 'es-model';

export function setMetadata(target: any, metadata: any, key?: string): void {
  if (key) {
    Reflect.defineMetadata(
      KEY,
      metadata,
      target,
      key,
    );
  } else {
    Reflect.defineMetadata(
      KEY,
      metadata,
      target,
    );
  }
}

export function getMetadata<T>(target: any, key?: string): T {
  if (key) {
    return Reflect.getMetadata(KEY, target, key);
  }

  return Reflect.getMetadata(KEY, target);
}

export function getPrimitiveType(objectType: Function, key: string): Function | null {
  return Reflect.getMetadata('design:type', objectType, key);
}
