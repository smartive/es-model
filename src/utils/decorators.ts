import { getMetadata, setMetadata } from './metadata';

export function generateObjectDecorator<T>(): (data: T) => ClassDecorator {
  return (data: T): ClassDecorator => (target: any) => setMetadata(target.prototype, data);
}

export function generatePropertyDecorator<T>(key: string): PropertyDecorator {
  return (data: T) => (target: Object, name: string) => {
    let mapping = getMetadata<{ [key: string]: T }>(target, key);
    if (!mapping) {
      mapping = {};
      setMetadata(target, mapping, key);
    }
    mapping[name] = data;
  };
}
