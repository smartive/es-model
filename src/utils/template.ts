import { Field, Fields, MappingMeta, SubMappingMeta } from '../models/Mapping';
import { getMetadata, getPrimitiveType } from './metadata';

export interface EsIndexTemplateMappings {
  [name: string]: EsIndexTemplateMapping;
}

export interface EsIndexTemplateMapping {
  dynamic?: 'strict';
  _source?: {
    enabled: boolean;
  };
  properties: EsIndexTemplateMappingFields;
}

export interface EsIndexTemplateMappingFields {
  [name: string]: EsIndexTemplateMappingField;
}

export interface EsIndexTemplateMappingField {
  type: string;
  include_in_parent?: boolean;
  search_analyzer?: string;
  analyzer?: string;
  format?: string;
  properties?: EsIndexTemplateMappingFields;
  fields?: EsIndexTemplateMappingFields;
}

export function buildMappings(types: Function[]): EsIndexTemplateMappings {
  const mappings: EsIndexTemplateMappings = {};

  for (const type of types) {
    const { name, mapping } = buildMapping(type.prototype);
    mappings[name || type.name.toLowerCase()] = mapping;
  }

  return mappings;
}

function buildMapping(type: Function): { name: string | undefined, mapping: EsIndexTemplateMapping } {
  const mapping: EsIndexTemplateMapping = {
    properties: {},
  };
  let name: string | undefined;
  const meta = getMetadata<MappingMeta>(type);

  if (meta) {
    name = meta.name;
    mapping._source = meta._source;
    mapping.dynamic = meta.dynamic;
  }
  
  const properties = getMetadata<Fields>(type, 'properties');
  if (properties) {
    mapping.properties = buildFields(properties, type);
  }

  return { name, mapping };
}

function buildFields(data: Fields, type: Function): EsIndexTemplateMappingFields {
  const fields: EsIndexTemplateMappingFields = {};
  for (const name of Object.keys(data)) {
    fields[name] = buildField(data[name], name, type);
  }
  return fields;
}

const PRIMITIVE_TYPES: { [key: string]: string } = {
  String: 'keyword',
  Boolean: 'boolean',
};

function buildField(field: Field, name: string, type: Function): EsIndexTemplateMappingField {
  let res: EsIndexTemplateMappingField = {
    type: 'keyword',
  };
  const primitiveType = getPrimitiveType(type, name);
  if (primitiveType && PRIMITIVE_TYPES[primitiveType.name]) {
    res.type = PRIMITIVE_TYPES[primitiveType.name];
  }

  if (field) {
    res = {
      ...field,
      ...res,
    };
    if (field.type) {
      if (field.type instanceof Function) {
        if (PRIMITIVE_TYPES[field.type.name]) {
          res.type = PRIMITIVE_TYPES[field.type.name];
        } else {
          res.type = 'object';

          const prototype = field.type.prototype;
      
          const fieldTypeMeta = getMetadata<SubMappingMeta>(prototype);
          if (fieldTypeMeta) {
            res = { ...res, ...fieldTypeMeta };
          }
      
          const properties = getMetadata<Fields>(prototype, 'properties');
          if (properties) {
            res.properties = buildFields(properties, prototype);
          }
          const fields = getMetadata<Fields>(prototype, 'fields');
          if (fields) {
            res.fields = buildFields(fields, prototype);
          }
        }
      } else if (field.type) {
        res.type = field.type;
      }
    }
  }

  return res;
}
