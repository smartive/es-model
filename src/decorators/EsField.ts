import { Field } from '../models/Mapping';
import { generatePropertyDecorator } from '../utils/decorators';

// tslint:disable-next-line:variable-name - This is a decorator
export const EsField = generatePropertyDecorator<Field>('fields');
