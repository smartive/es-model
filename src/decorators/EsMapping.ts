import { MappingMeta } from '../models/Mapping';
import { generateObjectDecorator } from '../utils/decorators';

// tslint:disable-next-line:variable-name - This is a decorator
export const EsMapping = generateObjectDecorator<MappingMeta>();
