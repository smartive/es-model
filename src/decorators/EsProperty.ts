import { Field } from '../models/Mapping';
import { generatePropertyDecorator } from '../utils/decorators';

// tslint:disable-next-line:variable-name - This is a decorator
export const EsProperty = generatePropertyDecorator<Field>('properties');
