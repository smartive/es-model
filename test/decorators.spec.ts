import 'reflect-metadata';

import { EsField } from '../src/decorators/EsField';
import { EsMapping } from '../src/decorators/EsMapping';
import { EsProperty } from '../src/decorators/EsProperty';
import { EsSubMapping } from '../src/decorators/EsSubMapping';
import { getMetadata } from '../src/utils/metadata';

describe('decorators', () => {
  describe('EsObject ', () => {
    it('generates metadata', () => {

      @EsSubMapping()
      class Obj {
        @EsProperty()
        aProperty: string;
      }

      @EsSubMapping({ type: 'nested' })
      class Nested {
        @EsProperty()
        aProperty: string;
      }

      @EsSubMapping({ type: 'keyword' })
      class WithFields {
        @EsField()
        keyword: string;
      }

      @EsMapping({
        dynamic: 'strict',
        _source: {
          enabled: true,
        },
      })
      class A {
        @EsProperty()
        keyword: string;

        @EsProperty({ type: String })
        keywordArray: string[];

        @EsProperty({ type: 'text' })
        text: string;
      
        @EsProperty({ type: Obj })
        b: Obj;

        @EsProperty({ type: Nested })
        b: Nested;

        @EsProperty({ type: WithFields })
        withFields: WithFields;
      }

      expect(getMetadata(A.prototype)).toMatchSnapshot();
      expect(getMetadata(A.prototype, 'properties')).toMatchSnapshot();
      expect(getMetadata(A.prototype, 'fields')).toMatchSnapshot();
      expect(getMetadata(Obj.prototype)).toMatchSnapshot();
      expect(getMetadata(Obj.prototype, 'properties')).toMatchSnapshot();
      expect(getMetadata(Obj.prototype, 'fields')).toMatchSnapshot();
      expect(getMetadata(Nested.prototype)).toMatchSnapshot();
      expect(getMetadata(Nested.prototype, 'properties')).toMatchSnapshot();
      expect(getMetadata(Nested.prototype, 'fields')).toMatchSnapshot();
      expect(getMetadata(WithFields.prototype)).toMatchSnapshot();
      expect(getMetadata(WithFields.prototype, 'properties')).toMatchSnapshot();
      expect(getMetadata(WithFields.prototype, 'fields')).toMatchSnapshot();
    });
  });
});
