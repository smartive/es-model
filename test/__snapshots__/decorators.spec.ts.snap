// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators EsObject  generates metadata 1`] = `
Object {
  "_source": Object {
    "enabled": true,
  },
  "dynamic": "strict",
}
`;

exports[`decorators EsObject  generates metadata 2`] = `
Object {
  "b": Object {
    "type": [Function],
  },
  "keyword": undefined,
  "keywordArray": Object {
    "type": [Function],
  },
  "text": Object {
    "type": "text",
  },
  "withFields": Object {
    "type": [Function],
  },
}
`;

exports[`decorators EsObject  generates metadata 3`] = `undefined`;

exports[`decorators EsObject  generates metadata 4`] = `undefined`;

exports[`decorators EsObject  generates metadata 5`] = `
Object {
  "aProperty": undefined,
}
`;

exports[`decorators EsObject  generates metadata 6`] = `undefined`;

exports[`decorators EsObject  generates metadata 7`] = `
Object {
  "type": "nested",
}
`;

exports[`decorators EsObject  generates metadata 8`] = `
Object {
  "aProperty": undefined,
}
`;

exports[`decorators EsObject  generates metadata 9`] = `undefined`;

exports[`decorators EsObject  generates metadata 10`] = `
Object {
  "type": "keyword",
}
`;

exports[`decorators EsObject  generates metadata 11`] = `undefined`;

exports[`decorators EsObject  generates metadata 12`] = `
Object {
  "keyword": undefined,
}
`;
